waku/v2/protocol/waku_rln_relay/rln_relay.nim

```diff
@@ -0,0 +1,421 @@
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+import
+  std/[algorithm, sequtils, strutils, tables, times, os, deques],
+  chronicles, options, chronos, stint,
+  confutils,
+  strutils,
+  web3, json,
+  web3/ethtypes,
+  eth/keys,
+  libp2p/protocols/pubsub/rpc/messages,
+  libp2p/protocols/pubsub/pubsub,
+  stew/results,
+  stew/[byteutils, arrayops]
+import
+  ./group_manager,
+  ./rln,
+  ./conversion_utils,
+  ./constants,
+  ./protocol_types,
+  ./protocol_metrics
+import
+  ../../utils/time,
+  ../waku_keystore
+
+logScope:
+  topics = "waku rln_relay"
+
+type WakuRlnConfig* = object
+  rlnRelayDynamic*: bool
+  rlnRelayPubsubTopic*: PubsubTopic
+  rlnRelayContentTopic*: ContentTopic
+  rlnRelayMembershipIndex*: Option[uint]
+  rlnRelayEthContractAddress*: string
+  rlnRelayEthClientAddress*: string
+  rlnRelayEthAccountPrivateKey*: string
+  rlnRelayEthAccountAddress*: string
+  rlnRelayCredPath*: string
+  rlnRelayCredentialsPassword*: string
+
+proc createMembershipList*(n: int): RlnRelayResult[(
+    seq[(string, string, string, string)], string
+  )] =
+  ## createMembershipList produces a sequence of identity credentials in the form of (identity trapdoor, identity nullifier, identity secret hash, id commitment) in the hexadecimal format
+  ## this proc also returns the root of a Merkle tree constructed out of the identity commitment keys of the generated list
+  ## the output of this proc is used to initialize a static group keys (to test waku-rln-relay in the off-chain mode)
+  ## Returns an error if it cannot create the membership list
+
+  # initialize a Merkle tree
+  let rlnInstance = createRLNInstance()
+  if rlnInstance.isErr():
+    return err("could not create rln instance: " & rlnInstance.error())
+  let rln = rlnInstance.get()
```

Comment:
I recommend passing this instance as a parameter (as it is created from the very beginning of the function). And create the instance one level up and check there the initialzation error.

Reply:
Addressed in 09da0a85